dill/logger.py

```diff
@@ -0,0 +1,153 @@
+        self.logger = logger
+    def process(self, msg, kwargs):
+        # A no-op override, as we don't have self.extra
+        return msg, kwargs
+    def log_trace(self, pickler, msg=None, *args, **kwargs):
+        """log_trace(self, pickler, msg=None, *args, reset=False, **kwargs)"""
+        reset = kwargs.pop('reset', False)
+        if msg is None:
+            # Initialization and clean-up.
+            if reset:
+                pickler._trace_depth = 0
+                pickler._size_stack = None
+            elif self.isEnabledFor(INFO_DETAIL):
+                pickler._trace_depth = 1
+                pickler._size_stack = []
+            return
+
+        elif getattr(pickler, '_trace_depth', None):
+            extra = kwargs.get('extra', {})
+            pushed = msg.startswith('#')
+            size = None
+            try:
+                size = pickler._file.tell()
+                size += pickler.framer.current_frame.tell()
+            except AttributeError:
+                pass
+            if size:
+                if not pushed:
+                    pickler._size_stack.append(size)
+                else:
+                    size -= pickler._size_stack.pop()
+                    extra['size'] = size
+            if pushed: pickler._trace_depth -= 1
+            extra['depth'] = pickler._trace_depth
+            kwargs['extra'] = extra
+            self.log(INFO_DETAIL, msg, *args, **kwargs)
+            if not pushed: pickler._trace_depth += 1
+
+        else:
+            self.info(msg, *args, **kwargs)
+
+class TraceFormatter(logging.Formatter):
+    """generates message prefix and suffix from record"""
+    def __init__(self, *args, **kwargs):
+        super(TraceFormatter, self).__init__(*args, **kwargs)
+        self.is_utf8 = locale.getpreferredencoding() == 'UTF-8'
+    def format(self, record):
+        fields = {'prefix': "", 'suffix': ""}
+        if getattr(record, 'depth', 0) > 0:
+            if record.msg.startswith("#"):
+                prefix = (record.depth-1)*u"│" + u"└ "
+            elif record.depth == 1:
+                prefix = u"┬ "
+            else:
+                prefix = (record.depth-2)*u"│" + u"├┬ "
+            if not self.is_utf8:
+                prefix = prefix[:-1].translate(TRANS_TABLE) + "- "
+            fields['prefix'] = prefix
+        if hasattr(record, 'size'):
+            # Show object size in human-redable form.
+            power = int(math.log(record.size, 2)) // 10
+            size = record.size >> power*10
+            fields['suffix'] = " [%d %sB]" % (size, "KMGTP"[power] + "i" if power else "")
+        record.__dict__.update(fields)
+        return super(TraceFormatter, self).format(record)
+
+
+logger = logging.getLogger('dill')
+adapter = TraceAdapter(logger)
+
+handler = logging.StreamHandler()
+logger.addHandler(handler)
+
+formatter = TraceFormatter("%(prefix)s%(message)s%(suffix)s")
+handler.setFormatter(formatter)
+
+
+def trace(boolean, detail=False):
```

Comment:
Why bother to have the option to turn off a more detailed trace?  Is there any value to a less detailed trace?  If not, then don't expose an option to turn off `detail`.  The new trace is nice... why wouldn't someone always use it?

Reply:
Sorry, I can't recall why I thought this should be optional. The only scenario I can think of in which it would be useful to not indent the trace in a tree style is for inspecting the trace of some (very) deeply nested object structure. But even then I don't think the indentation would obscure it much... Maybe it's better to just drop the option (and the `INFO_DETAIL` log level with it).

Reply:
I don't see any reason for switching off `detail`, so if you don't either, then remove the option.

Reply:
Option removed.